### PR TITLE
feature(CB2-3950): Create a Test Stations Service and Store to Enable Auto-Completion

### PR DESCRIPTION
**Pull request by tahayazici**
## Amend Visit Section

- Create a test stations store with relevant action, reducers, selectors & effects.
- Create a simple test stations service to manage API calls.
- Amend the visit section's template to make relevant fields editable.
- Add the test station options into the dynamic field component.
- Update the auto-complete component to work with the new API data.

[CB2-3950](https://dvsa.atlassian.net/browse/CB2-3950)

## Checklist

- [x] Branch is rebased against the latest develop/common
- [ ] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [x] Code and UI has been tested manually after the additional changes
- [x] PR title includes the JIRA ticket number
- [x] Squashed commits contain the JIRA ticket number
- [ ] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

File: src/app/forms/components/dynamic-form-field/dynamic-form-field.component.ts [hunk 1]
```diff
@@ -13,31 +15,43 @@ import { map, mergeMap, Observable, of } from 'rxjs';
...
+      );
+    } else if (meta?.referenceData) {
+      return this.referenceDataService
+        .getAll$((this.control?.value.meta.referenceData ?? '') as ReferenceDataResourceType)
+        .pipe(
+          map(options => options.map(option => ({ value: option.resourceKey, label: option.description })))
+        );
+    } else {
+      return of(this.control?.value.meta.options as FormNodeOption<string | number | boolean>[] ?? []);
+    }
```
[the remark this answers is not in the document]
Comment (mcadenas-bjss) [reply src/app/forms/components/dynamic-form-field/dynamic-form-field.component.ts:43]: Could map the data coming out of the API into the common reference data model so everything is uniform or map it from the selector

[the remark this answers is not in the document]
Comment (tahayazici) [reply src/app/forms/components/dynamic-form-field/dynamic-form-field.component.ts]: Amazing suggestion! I agree that the mapping is complex and not the dynamic form field's responsibility, so it does warrant abstracting into the services.

Suggestion applied.

**Review by mcadenas-bjss — APPROVED** [review]

**Review by Ian-Hodges — DISMISSED** [review]

File: src/app/services/test-stations/test-stations.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+      // Check for correct requests: should have made one request to search from expected URL
+      const req = httpTestingController.expectOne(`${environment.VTM_API_URI}/test-stations/`);
+      expect(req.request.method).toEqual('GET');
+
+      // Provide each request with a mock response
+      req.flush(expectedResult);
+    });
+
+    it('should handle errors', () => {
```
Comment (gjulien-bjss) [inline src/app/services/test-stations/test-stations.service.spec.ts:45]: This test and the one one line 72 seem to not be doing what they are supposed to. Why would you expect to have a test station in the data if the server responds with a 500? I think you need to use the `done()` function inside the subscribe. Also these unit tests are both failing on the branch, if you look in the `checks` tab on the PR.

**Review by gjulien-bjss — CHANGES_REQUESTED** [review]

File: src/app/services/test-stations/test-stations.service.spec.ts (hunk 2 above)
Comment (tahayazici) [reply src/app/services/test-stations/test-stations.service.spec.ts]: Suggestion applied. Good spot - thanks!

**Review by mcadenas-bjss — APPROVED** [review]

**Review by gjulien-bjss — APPROVED** [review]

**Pull request merged**